src/System.Linq.Expressions/src/System/Linq/Expressions/TypeBinaryExpression.cs

```diff
@@ -79,11 +79,17 @@ internal Expression ReduceTypeEqual()
                 return ReduceConstantTypeEqual();
             }
 
-            // If the operand type is a sealed reference type or a nullable
-            // type, it will match if value is not null
-            if (cType.GetTypeInfo().IsSealed && (cType == _typeOperand))
+            // If the expression type is a sealed reference type or a nullable
+            // type, it will match if the value is not null and the type operand
+            // either matches or one is a nullable type while the other is its
+            // type argument (T to the other's T?).
+            if (cType.GetTypeInfo().IsSealed)
             {
-                if (cType.IsNullableType())
+                if (cType.GetNonNullableType() != _typeOperand.GetNonNullableType())
+                {
+                    return Expression.Constant(false);
```

Comment:
It seems we're no longer evaluating `Expression` for its side-effects in the cases where we return a constant now.